File: backend/api/documents_api.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from typing import List, Optional
from datetime import datetime
import json

from app.database import get_db
from app.models import Document, User
from app.schemas import DocumentCreate, DocumentUpdate, DocumentResponse, DocumentApproval
from app import auth

# Используем правильную функцию аутентификации
get_current_user = auth.get_current_user

router = APIRouter(prefix="/documents", tags=["documents"])

@router.get("/", response_model=List[DocumentResponse])
async def get_documents(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
    document_type: Optional[str] = None,
    status: Optional[str] = None
):
    """Получить список документов пользователя"""
    query = db.query(Document).filter(Document.created_by == current_user.id)
    
    if document_type:
        query = query.filter(Document.document_type == document_type)
    
    if status:
        query = query.filter(Document.status == status)
    
    documents = query.order_by(Document.created_at.desc()).all()
    return documents

@router.get("/pending-approvals", response_model=List[DocumentResponse])
async def get_pending_approvals(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Получить документы, ожидающие согласования руководителем"""
    # Проверяем, является ли пользователь руководителем
    if current_user.role not in ['admin', 'manager']:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Недостаточно прав для просмотра документов на согласовании"
        )
    
    # Получаем документы, где текущий пользователь является согласующим
    documents = db.query(Document).filter(
        Document.approver_id == current_user.id,
        Document.status == "pending"
    ).order_by(Document.created_at.desc()).all()
    
    return documents

@router.post("/", response_model=DocumentResponse)
async def create_document(
    document: DocumentCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Создать новый документ"""
    # Определяем согласующего (руководителя подразделения)
    approver = db.query(User).filter(
        User.department == current_user.department,
        User.role.in_(['admin', 'manager'])
    ).first()
    
    if not approver:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Не найден руководитель подразделения для согласования"
        )
    
    # Создаем документ
    db_document = Document(
        document_type=document.document_type,
        title=document.title,
        data=json.dumps(document.data),
        status="pending",
        created_by=current_user.id,
        approver_id=approver.id,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow()
    )
    
    db.add(db_document)
    db.commit()
    db.refresh(db_document)
    
    return db_document

@router.get("/{document_id}", response_model=DocumentResponse)
async def get_document(
    document_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Получить документ по ID"""
    document = db.query(Document).filter(Document.id == document_id).first()
    
    if not document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Документ не найден"
        )
    
    # Проверяем права доступа
    if document.created_by != current_user.id and document.approver_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Нет прав для просмотра этого документа"
        )
    
    return document

@router.put("/{document_id}/approve", response_model=DocumentResponse)
async def approve_document(
    document_id: int,
    approval: DocumentApproval,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Согласовать или отклонить документ"""
    document = db.query(Document).filter(Document.id == document_id).first()
    
    if not document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Документ не найден"
        )
    
    # Проверяем, что пользователь является согласующим
    if document.approver_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Нет прав для согласования этого документа"
        )
    
    # Проверяем, что документ в статусе ожидания
    if document.status != "pending":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Документ уже обработан"
        )
    
    # Обновляем статус документа
    document.status = "approved" if approval.approved else "rejected"
    document.approval_comment = approval.comment
    document.updated_at = datetime.utcnow()
    
    db.commit()
    db.refresh(document)
    
    return document

@router.put("/{document_id}", response_model=DocumentResponse)
async def update_document(
    document_id: int,
    document_update: DocumentUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Обновить документ"""
    document = db.query(Document).filter(Document.id == document_id).first()
    
    if not document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Документ не найден"
        )
    
    # Проверяем, что пользователь является автором документа
    if document.created_by != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Нет прав для редактирования этого документа"
        )
    
    # Проверяем, что документ можно редактировать
    if document.status != "draft":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Можно редактировать только черновики"
        )
    
    # Обновляем документ
    if document_update.title:
        document.title = document_update.title
    if document_update.data:
        document.data = json.dumps(document_update.data)
    
    document.updated_at = datetime.utcnow()
    
    db.commit()
    db.refresh(document)
    
    return document

@router.delete("/{document_id}")
async def delete_document(
    document_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Удалить документ"""
    document = db.query(Document).filter(Document.id == document_id).first()
    
    if not document:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Документ не найден"
        )
    
    # Проверяем, что пользователь является автором документа
    if document.created_by != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Нет прав для удаления этого документа"
        )
    
    # Проверяем, что документ можно удалить
    if document.status != "draft":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Можно удалить только черновики"
        )
    
    db.delete(document)
    db.commit()
    
    return {"message": "Документ удален"}

@router.get("/types/available")
async def get_available_document_types(
    current_user: User = Depends(get_current_user)
):
    """Получить доступные типы документов для пользователя"""
    base_types = [
        {
            "type": "vacation",
            "title": "Заявление на отпуск",
            "icon": "calendar",
            "color": "blue",
            "available_for": ["user", "manager", "admin"]
        },
        {
            "type": "payment",
            "title": "Заявление на выплаты",
            "icon": "dollar",
            "color": "green",
            "available_for": ["user", "manager", "admin"]
        },
        {
            "type": "resignation",
            "title": "Заявление на увольнение",
            "icon": "logout",
            "color": "red",
            "available_for": ["user", "manager", "admin"]
        },
        {
            "type": "instruction",
            "title": "Служебная инструкция",
            "icon": "file-protect",
            "color": "purple",
            "available_for": ["user", "manager", "admin"],
            "is_printable": True
        }
    ]
    
    # Фильтруем типы документов по роли пользователя
    available_types = [
        doc_type for doc_type in base_types 
        if current_user.role in doc_type["available_for"]
    ]
    
    return available_types

@router.get("/statistics")
async def get_document_statistics(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Получить статистику по документам"""
    # Документы пользователя
    my_documents = db.query(Document).filter(Document.created_by == current_user.id).all()
    
    # Документы на согласовании (если пользователь руководитель)
    pending_approvals = []
    if current_user.role in ['admin', 'manager']:
        pending_approvals = db.query(Document).filter(
            Document.approver_id == current_user.id,
            Document.status == "pending"
        ).all()
    
    # Статистика
    stats = {
        "total_my_documents": len(my_documents),
        "pending_my_documents": len([d for d in my_documents if d.status == "pending"]),
        "approved_my_documents": len([d for d in my_documents if d.status == "approved"]),
        "rejected_my_documents": len([d for d in my_documents if d.status == "rejected"]),
        "pending_approvals": len(pending_approvals),
        "by_type": {}
    }
    
    # Статистика по типам документов
    for doc in my_documents:
        doc_type = doc.document_type
        if doc_type not in stats["by_type"]:
            stats["by_type"][doc_type] = {
                "total": 0,
                "pending": 0,
                "approved": 0,
                "rejected": 0
            }
        
        stats["by_type"][doc_type]["total"] += 1
        stats["by_type"][doc_type][doc.status] += 1
    
    return stats 

File: frontend/src/api/documentsApi.ts
// Простой API клиент для документов
const apiClient = {
  get: async (url: string) => {
    const token = localStorage.getItem('token');
    const response = await fetch(`/api${url}`, {
      headers: { Authorization: `Bearer ${token}` }
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  },
  
  post: async (url: string, data: any) => {
    const token = localStorage.getItem('token');
    const response = await fetch(`/api${url}`, {
      method: 'POST',
      headers: { 
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}` 
      },
      body: JSON.stringify(data)
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  },
  
  put: async (url: string, data: any) => {
    const token = localStorage.getItem('token');
    const response = await fetch(`/api${url}`, {
      method: 'PUT',
      headers: { 
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}` 
      },
      body: JSON.stringify(data)
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  },
  
  delete: async (url: string) => {
    const token = localStorage.getItem('token');
    const response = await fetch(`/api${url}`, {
      method: 'DELETE',
      headers: { Authorization: `Bearer ${token}` }
    });
    if (!response.ok) {
      throw new Error(`HTTP error! status: ${response.status}`);
    }
    return response.json();
  }
};

export interface Document {
  id: number;
  document_type: string;
  title: string;
  data: any;
  status: 'draft' | 'pending' | 'approved' | 'rejected';
  created_by: number;
  approver_id: number;
  created_at: string;
  updated_at: string;
  approval_comment?: string;
}

export interface DocumentCreate {
  document_type: string;
  title: string;
  data: any;
}

export interface DocumentUpdate {
  title?: string;
  data?: any;
}

export interface DocumentApproval {
  approved: boolean;
  comment?: string;
}

export interface DocumentType {
  type: string;
  title: string;
  icon: string;
  color: string;
  available_for: string[];
  is_printable?: boolean;
}

export interface DocumentStatistics {
  total_my_documents: number;
  pending_my_documents: number;
  approved_my_documents: number;
  rejected_my_documents: number;
  pending_approvals: number;
  by_type: Record<string, {
    total: number;
    pending: number;
    approved: number;
    rejected: number;
  }>;
}

// Получить список документов пользователя
export const getDocuments = async (
  documentType?: string,
  status?: string
): Promise<Document[]> => {
  const params = new URLSearchParams();
  if (documentType) params.append('document_type', documentType);
  if (status) params.append('status', status);
  
  const response = await apiClient.get(`/documents/?${params.toString()}`);
  return response;
};

// Получить документы, ожидающие согласования
export const getPendingApprovals = async (): Promise<Document[]> => {
  const response = await apiClient.get('/documents/pending-approvals');
  return response;
};

// Создать новый документ
export const createDocument = async (document: DocumentCreate): Promise<Document> => {
  const response = await apiClient.post('/documents/', document);
  return response;
};

// Получить документ по ID
export const getDocument = async (documentId: number): Promise<Document> => {
  const response = await apiClient.get(`/documents/${documentId}`);
  return response;
};

// Согласовать или отклонить документ
export const approveDocument = async (
  documentId: number,
  approval: DocumentApproval
): Promise<Document> => {
  const response = await apiClient.put(`/documents/${documentId}/approve`, approval);
  return response;
};

// Обновить документ
export const updateDocument = async (
  documentId: number,
  update: DocumentUpdate
): Promise<Document> => {
  const response = await apiClient.put(`/documents/${documentId}`, update);
  return response;
};

// Удалить документ
export const deleteDocument = async (documentId: number): Promise<void> => {
  await apiClient.delete(`/documents/${documentId}`);
};

// Получить доступные типы документов
export const getAvailableDocumentTypes = async (): Promise<DocumentType[]> => {
  const response = await apiClient.get('/documents/types/available');
  return response;
};

// Получить статистику по документам
export const getDocumentStatistics = async (): Promise<DocumentStatistics> => {
  const response = await apiClient.get('/documents/statistics');
  return response;
}; 